whatshap/polyphase/algorithm.py

```diff
@@ -0,0 +1,257 @@
+        joblist = list(zip(range(len(block_starts)), block_starts[:-1], block_starts[1:]))
+        joblist.sort(key=lambda x: x[1] - x[2])
+
+        with Pool(processes=param.threads) as pool:
+            process_results = [
+                pool.apply_async(
+                    phase_single_block_mt,
+                    (
+                        allele_matrix,
+                        block_id,
+                        start,
+                        end,
+                        genotype_list[start:end],
+                        param,
+                        timers,
+                        job_id,
+                        num_blocks,
+                    ),
+                )
+                for job_id, (block_id, start, end) in enumerate(joblist)
+            ]
+            # collect all blockwise results
+            blockwise_results = [res.get() for res in process_results]
+            results = sorted(blockwise_results, key=lambda x: x.block_id)
+
+        timers.stop("phase_blocks")
+
+    # Aggregate blockwise results
+    return aggregate_results(results, ploidy)
+
+
+def phase_single_block(block_id, allele_matrix, genotype_slice, param, timers, quiet=False):
+    """
+    Takes as input data the reads from a single (pre-computed) block and the genotypes for all
+    variants inside the block. Runs a three-phase algorithm to compute a phasing for this isolated
+    block. Input are four objects:
+
+    allele_matrix -- Input reads
+    genotype_slice -- Genotype (as dictionary) for every position
+    param -- Object containing phasing parameters
+    timers -- Timer object to measure time
+    """
+
+    block_num_vars = allele_matrix.getNumPositions()
+
+    # Check for singleton blocks and handle them differently (for efficiency reasons)
+    if block_num_vars == 1:
+
+        # construct trivial solution for singleton blocks, by using the genotype as phasing
+        g = genotype_slice[0]
+        clusts = [[i for i, r in enumerate(allele_matrix) if r and r[0][1] == a] for a in g]
+        paths = [sorted(list(chain(*[[i] * g[a] for i, a in enumerate(g)])))]
+        haps = sorted(list(chain(*[[[a]] * g[a] for a in g])))
+        return PolyphaseBlockResult(block_id, clusts, paths, [0], [[0]] * param.ploidy, haps)
+
+    # Block is non-singleton here, so run the normal routine
+    # Phase I: Cluster Editing
+
+    # Compute similarity values for all read pairs
+    timers.start("read_scoring")
+    logger.debug("Computing similarities for read pairs ..")
+    sim = scoreReadset(allele_matrix, param.min_overlap, param.ploidy, 0.07)
+    timers.stop("read_scoring")
+
+    # Run cluster editing
+    timers.start("clustering")
+    logger.debug(
+        "Solving cluster editing instance with {} nodes and {} edges ..".format(
+            len(allele_matrix), len(sim)
+        )
+    )
+    solver = ClusterEditingSolver(sim, param.ce_bundle_edges)
+    clustering = solver.run()
+    del solver
```

Comment:
I removed a deletion statement at another function and indeed letting a cython object move out of scope is the same (memory-consumption-wise) as deleting it manually. In this case I would leave it, because I think it's not really worth creating another function for this short code.